### PR TITLE
Make sure autocompleted values make it into the block's saved content

### PR DESCRIPTION
**Pull request by notnownikki**
## Description

Potential fix for https://github.com/WordPress/gutenberg/issues/7444

When an autocompleter does an insert-at-caret, the block isn't notified of the change,
and so the content doesn't get updated unless the user carries on typing.

This change dispatches an input event, which the editable component listens for and updates the block's attributes.

## How has this been tested?

Carry out the steps shown in https://github.com/WordPress/gutenberg/issues/7444
and the username should be saved correctly.

## Types of changes
Bug fix

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation.

### CODE REVIEW

**Comment by notnownikki** [comment]
This is the first time I've looked at this code, so I'm suggesting this as a potential approach. If it's acceptable, I'll write tests :)

File: components/autocomplete/index.js
Comment (gziolo) [inline components/autocomplete/index.js]: Should we default to `noop` instead? Can we update docs to reflect this new change introduced?

Comment (notnownikki) [reply components/autocomplete/index.js]: Sounds good to me, at this point I just want confirmation that this is the right type of fix, and I'll add a default noop and write tests if it is :)

**Review by brandonpayton — CHANGES_REQUESTED** [review]
Hi @notnownikki, thanks for working on this! I was thinking I'd need to look into this issue this morning, and it was a pleasant surprise to see you already had.

I'm thinking it might be better if we manually dispatch an `input` event on the contenteditable element since we already listen for an `input` event and it represents what is occurring and wouldn't require any interface change.

Maybe something like this at the end of `insertCompletion`:

```javascript
// IE11 doesn't provide an InputEvent constructor.
const inputEvent = typeof InputEvent !== 'undefined' ? new InputEvent() : (() => {
  const e = document.createEvent( 'UIEvent' );
  e.initEvent( 'input', true /* bubbles */, false /* cancelable */ );
  return e;
})();
range.commonAncestorContainer.closest( '[contenteditable=true]' ).dispatch( inputEvent );
```

What do you think?

**Comment by notnownikki** [comment]
@brandonpayton that's a much cleaner solution! I've pushed up a new change that dispatches in input event, and it all seems to work fine :)

**Review by mcsf — COMMENTED** [review]
This fix looks good to me, though at the moment my IE11 virtual machine is not cooperating.

File: components/autocomplete/index.js
Comment (notnownikki) [reply components/autocomplete/index.js]: I was following @brandonpayton's lead on this... but looking at it, for consistency, shouldn't it be an InputEvent?

Comment (mcsf) [reply components/autocomplete/index.js]: Perhaps. :) Since this is an IE quirk, I don't know. I brought up `Event` because the typical examples out there on how to "deal with IE" use it instead of other event types.

Comment (mcsf) [reply components/autocomplete/index.js]: Actually, `InputEvent` is not supported at all, even with the fallback interface. It has to be either `Event` or `UIEvent`.

**Review by mcsf — APPROVED** [review]
I don't feel strongly enough about the `Event`/`UIEvent` choice and don't mean to block this PR — so, if this PR also correctly fixes the problem in IE11, feel free to merge it!

File: components/autocomplete/index.js
Comment (brandonpayton) [reply components/autocomplete/index.js]: I think we would be fine using `Event`, but the reason is that the `InputEvent` interface is based on `UIEvent` in the spec:
https://w3c.github.io/uievents/#idl-inputevent

**Comment by notnownikki** [comment]
Tested in IE11, works fine!

**Review by brandonpayton — APPROVED** [review]
I tested in IE11, and the fix is effective to address #7444. However, I am seeing our manually dispatched `input` event trigger another display of the completion menu. This also occurs in Firefox.
reopened-completion-menu

If I hack`insertCompletion` to insert a space after a completion, the menu is not shown after the completion is inserted. This makes some sense as the originally inserted text would indeed trigger the completer menu if it was input by the user.

Once the autocomplete token PR #6577 is merged (hopefully later today), this will no longer occur because the cursor is placed outside completion tokens after insertion.

I am happy with this change but think we should wait to merge after #6577.

**Comment by notnownikki** [comment]
@brandonpayton thanks! I'm happy to wait :)

**Comment by gziolo** [comment]
@brandonpayton - should we still wait for #6577?

**Comment by brandonpayton** [comment]
@gziolo, I don't want to wait, but there is a bug with this solution if we do not have separation because the input event triggers `Autocomplete` to [redisplay the menu](https://github.com/WordPress/gutenberg/pull/7486#pullrequestreview-132463737) since the cursor is still at the end of a pattern matching the user autocompleter.

A temporary workaround might be to insert a non-breaking zero-width space following a completion. It might introduce a surprising stop when moving the cursor with arrows, but that seems more tolerable than this bug. What do you think?

**Comment by brandonpayton** [comment]
@gziolo I realized there is another solution to the menu-redisplay issue that doesn't require waiting on #6577. I pushed a commit with that fix. If the change looks good to you, I think this is ready to merge.

**Comment by gziolo** [comment]
Yes, works for me. I can see user handle properly published 🎉

**Pull request merged**
